log4j-core/src/main/java/org/apache/logging/log4j/core/appender/rolling/action/CompressActionFactory.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.core.appender.rolling.action;
+
+import java.nio.file.Path;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.logging.log4j.core.util.FileUtils;
+import org.apache.logging.log4j.core.util.NetUtils;
+import org.jspecify.annotations.NullMarked;
+
+/**
+ * A factory to produce actions that compress log files.
+ */
+@NullMarked
+public interface CompressActionFactory {
+
+    /**
+     * Option to specify the compression level.
+     * <p>
+     *     The value must be an {@code int} between -1 (default) and 9.
+     * </p>
+     */
+    String COMPRESSION_LEVEL = "compressionLevel";
```

Comment:
IMHO,

1. This is not the correct place for this configuration option name.
2. We shouldn't pass configurations in a `Map<String, ?>`, but in a well-typed object/record instead. Can't we adapt the code in that direction instead?

Reply:
Ad. 1
This option is hardcoded in the configuration of each `RolloverStrategy`, which currently provides a compression level as only way to configure configuration. This seems as a good intersection point between `RolloverStrategy` (which uses `CompressActionFactory`) and two `CompressActionFactory` implementations, which actually use the parameter. What better place would you suggest?

Ad. 2
Would you prefer `Map<String, String>`? There are multiple approaches to provide a generic way to pass parameters for an _a priori_ unknown `CompressActionFactory`:

- a `Map<String, String>`, which does not generate any `ClassCastException`s, but might generate parsing errors if a parameter can not be parsed (e.g. the value of `compressionLevel` can not be parsed as string).
- a `Map<String, ?>, which does not suffer from parsing errors, but causes `ClassCastException`s. :cry: 
- a marker interface, so that each algorithm can provide its own well-typed object/record. This will cause `IllegalArgumentException`s if the params for factory A are passed to factory B.
- parameterize `CompressActionFactory` with the type of options it uses. Everything works nicely in the code, but we are mostly using reflection.

I don't have a preferred strategy, but the `Map` approaches have the advantage that we can provide the user with a way to fill in a `Map`, pass it to Commons Compress and let it worry about interpreting them.
With this approach, if Commons Compress adds additional options to [GzipParameters](https://javadoc.io/doc/org.apache.commons/commons-compress/latest/org/apache/commons/compress/compressors/gzip/GzipParameters.html), users could immediately profit from them (assuming we add code to convert `Map` to `GzipParameters`.

Reply:
I see. Given compression is activated based on the extension of the archived file name, can we instead

1. Remove/Deprecate the hardcoded `compressionLevel` attribute, and
2. Create a new `Compression` element holding arbitrary key-value pairs?

Example snippet:

```xml
<RollingFile name="FILE"
             fileName="/var/log/app.log"
             filePattern="/var/log/app.log.%i.gz"> 
  <JsonTemplateLayout/>
  <DefaultRolloverStrategy max="15"/> 
  <Policies>
    <CronTriggeringPolicy schedule="0 0 0 * * ?"/> 
    <SizeBasedTriggeringPolicy size="100k"/> 
  </Policies>
  <Compression>
    <level>10</level>
    <algorithm>gzip</algorithm>
    <bufferSize>8192</bufferSize>
  </Compression>
</RollingFile>
```

Reply:
I created #2950 test for user support for this idea.

In the meantime I have implemented the removal of `compressionLevel` in https://github.com/apache/logging-log4j2/pull/2921/commits/806a0be16e43dff72b5f0645d704b594d30a8bbb